cmd/hegel/root_cmd.go

```diff
@@ -105,52 +120,110 @@ func NewRootCommand() (*RootCommand, error) {
 	return rootCmd, nil
 }
 
-// PreRun satisfies cobra.Command.PreRunE and unmarshalls. Its responsible for populating rc.Opts.
-func (rc *RootCommand) PreRun(*cobra.Command, []string) error {
-	if err := rc.Vpr.Unmarshal(&rc.Opts); err != nil {
+// PreRun satisfies cobra.Command.PreRunE and unmarshalls. Its responsible for populating c.Opts.
+func (c *RootCommand) PreRun(*cobra.Command, []string) error {
+	if err := c.vpr.Unmarshal(&c.Opts); err != nil {
 		return err
 	}
 
-	return rc.validateOpts()
+	return c.validateOpts()
 }
 
-// Run executes Hegel. Its temporarily unimplemented.
-func (rc *RootCommand) Run(*cobra.Command, []string) error {
-	return nil
+// Run executes Hegel.
+func (c *RootCommand) Run(cmd *cobra.Command, _ []string) error {
+	logger, err := log.Init("github.com/tinkerbell/hegel")
+	if err != nil {
+		return fmt.Errorf("initialize logger: %w", err)
+	}
+	defer logger.Close()
+	metrics.Init(logger)
+
+	logger.Package("main").With("opts", fmt.Sprintf("%+v", c.Opts)).Info("root command options")
+
+	ctx, otelShutdown := otelinit.InitOpenTelemetry(cmd.Context(), "hegel")
+	defer otelShutdown(ctx)
+
+	metrics.State.Set(metrics.Initializing)
+
+	hardwareClient, err := hardware.NewClient(c.Opts.Facility, c.Opts.GetDataModel())
+	if err != nil {
+		return fmt.Errorf("create client: %w", err)
+	}
+
+	grpcServer := grpcserver.NewServer(logger, hardwareClient)
+
+	ctx, cancel := signals.ContextCanceller(cmd.Context(), syscall.SIGINT, syscall.SIGTERM)
+	var routines run.Group
+
+	routines.Add(
```

Comment:
would it make more sense to use `g, ctx := errgroup.WithContext(...)`. TIL about oklog/run and gotta say not a fan of the api taking 2 funcs, this seems more go like to me:

```go
ctx, cancel := signals.ContextCanceller(cmd.Context(), syscall.SIGINT, syscall.SIGTERM)
defer cancel()

g, ctx := errgroup.WithContext(ctx)
g.Go(func() error { return httpserver.Serve(...) })
g.Go(func() error { return grpcserver.Serve(...) })
return g.Wait()
```
The whole 2 func thing, one for error handling is mostly the weird part :). Its also nice that we can just defer the cancel asap, although I'm pretty sure we can do that with oklog/run too.

Reply:
`errgroup` is a good alternative with the tradeoff that it expects everything to work with contexts which isn't the case. Many libraries expose methods for performing shutdown. A good example of that is the `net/http.Server` construct that doesn't actually use contexts for shutdown, it exposes a `Shutdown()` (that accepts a context for timeout type behavior) and `Close()`. 

I don't think 2 funcs is a bad thing. Its an API that delegates the behavior rather than assumes its a context that needs cancelling.

Of course, both can be made to work in the same way as the other so I feel like this is 6 of 1, half a dozen of the other. I'm good to use either.

Reply:
right, but at this point I think its fair to say "context is the way". Hmm actually I'm now seeing that http servers don't have a context for shutting down and that seems annoying. Meh either way then too. Since this is already written then lets keep it. NM.

Reply:
We can always change later. I'm not opposed to rallying around errgroup soon.

Reply:
Yeah, agree. I think I do have a branch laying around somewhere for tink I think that used errgroup. I'll compare with `run` whenever I get to it.